lib/utils/createConfig.js

```diff
@@ -1,21 +1,20 @@
 'use strict';
 
-const path = require('path');
-const isAbsoluteUrl = require('is-absolute-url');
+/* eslint-disable
+  no-undefined
+*/
+
 const defaultTo = require('./defaultTo');
 
 function createConfig(config, argv, { port }) {
   const firstWpOpt = Array.isArray(config) ? config[0] : config;
   const options = firstWpOpt.devServer || {};
 
-  // This updates both config and firstWpOpt
-  firstWpOpt.mode = defaultTo(firstWpOpt.mode, 'development');
-
   if (argv.bonjour) {
     options.bonjour = true;
   }
 
-  if (argv.host && (argv.host !== 'localhost' || !options.host)) {
+  if (argv.host) {
```

Comment:
Why we remove `&& (argv.host !== 'localhost' || !options.host)`?

Reply:
We do not have to make this change now, but we need to carefully rethink how we handle the options: `host`, `port`, and `socket`, since these are relevant for the `listen` method.

The logic in `&& (argv.host !== 'localhost' || !options.host)` is made difficult by the fact that `localhost` is the CLI default, so we can't tell if the user explicitly asked for `localhost` or if it was set by default:

https://github.com/webpack/webpack-dev-server/blob/eab6acd2972e39c8b2d83f7a6614d40b8e14bf63/bin/options.js#L175

My proposal is:

**Don't set default values for the CLI options host or port**. I have done this in `webpack-cli serve`:
https://github.com/webpack/webpack-cli/blob/44d0fb01f47852b4dfc7423b8e1dfa6c51872e55/packages/serve/flags.js

**Have a clear hierarchy for which value gets used**. The order of precedence should be:
1. CLI value
2. webpackConfig.devServer value
3. default value

You can see I have implemented this order of precedence in these lines:
https://github.com/webpack/webpack-cli/pull/1011/files#diff-24921f08fe12bbbb2c25f5ed41d757b4R19-R23

Reply:
Oh, i think we should keep https://github.com/webpack/webpack-cli/blob/44d0fb01f47852b4dfc7423b8e1dfa6c51872e55/packages/serve/flags.js on our side, because when we want add cli option it is require change something on webpack-cli side, it is not good

Reply:
i.e. webpack-cli will work with `webpack-dev-server@3`?

Reply:
@Loonride `webpack-cli` should not have `webpack-dev-server` in deps, they should use `peerDependenciesMeta`